### ISSUE TITLE
uacctd does not work with non-L2 interface

### ISSUE THREAD

**Issue by aleksandrgilfanov** [pmacct/pmacct#387]
Hello, @paololucente. Is uacctd supposed to work with non-L2 interfaces?

I am trying it with hdlc-ip interfaces. It is working flawlessly when I build it with --disable-L2. And its not working at all, when I use --enable-l2.

Quick investigation through source code shows that eth_handler is always used when (HAVE_L2) is defined.

So, is it correct behaviour? There is no possibility to use both L2 and non-L2 interfaces on the same system? (with only one binary) 

**Version**
```
admin@edge# uacctd -V
Linux NetFilter NFLOG Accounting Daemon, uacctd 1.7.4-git (20191126-01+c6)

Arguments:
 '--build=x86_64-linux' '--host=x86_64-numa-linux' '--target=x86_64-numa-linux' '--prefix=/usr' '--exec_prefix=/usr' '--bindir=/usr/bin' '--sbindir=/usr/sbin' '--libexecdir=/usr/libexec' '--datadir=/usr/share' '--sysconfdir=/etc' '--sharedstatedir=/usr/com' '--localstatedir=/var' '--libdir=/usr/lib' '--includedir=/usr/include' '--oldincludedir=/usr/include' '--infodir=/usr/share/info' '--mandir=/usr/share/man' '--with-libtool-sysroot=/home/build/edge-oe/tmp/sysroots/amd64-numa-linux' '--enable-largefile' '--with-pcap-includes=/home/build/edge-oe/tmp/sysroots/amd64-numa-linux/usr/include' '--enable-nflog' '--enable-threads' '--enable-64bit''--enable-ipv6' 'build_alias=x86_64-linux' 'host_alias=x86_64-numa-linux' 'target_alias=x86_64-numa-linux' 'CC=ccache x86_64-numa-linux-gcc -march=x86-64 -mtune=generic --sysroot=/home/build/edge-oe/tmp/sysroots/amd64-numa-linux' 'CFLAGS=-pipe -O2 -fomit-frame-pointer -frename-registers -flto=8 -fno-lto -g -O0' 'LDFLAGS=-Wl,-O1 -Wl,--hash-style=gnu -Wl,-z,relro,-z,now -Wl,--build-id=none' 'CPPFLAGS=' 'CPP=x86_64-numa-linux-gcc -E --sysroot=/home/build/edge-oe/tmp/sysroots/amd64-numa-linux' '--enable-l2' '--enable-traffic-bins' '--enable-bgp-bins' '--enable-bmp-bins' '--enable-st-bins'

Libs:
libpcap version 1.7.4
netfilter_log

System:
Linux 4.4.26 #numa-edge SMP 2018-07-11 12:00:00 +0400 x86_64

Compiler:
gcc 4.9.2

For suggestions, critics, bugs, contact me: Paolo Lucente <paolo@pmacct.net>.
```

**Comment by paololucente** [comment]
Hi @aleksandrgilfanov ,

Thank you for reporting this. I spotted the line of code you mention (that makes all pass through the eth_handler()) if --enable-l2 is specified. Could you please send me a brief trace in libpcap format of hdlc-ip traffic by unicast emal? You can use tcpdump a-la: "tcpdump -i nflog:<nflog group> -w <output file>".

Should you have also some ethernet traffic that you can throw in the mix with hdlc-ip over the same nflog group, that would be awesome.

Paolo

**Comment by aleksandrgilfanov** [comment]
Actually, I fixed this problem with such patch:
```
diff --git a/src/uacctd.c b/src/uacctd.c
index ee67a22..af10206 100644
--- a/src/uacctd.c
+++ b/src/uacctd.c
@@ -56,9 +56,12 @@ static int nflog_incoming(struct nflog_g_handle *gh, struct nfgenmsg *nfmsg,
   struct pcap_pkthdr hdr;
   char *pkt = NULL;
   ssize_t pkt_len = nflog_get_payload(nfa, &pkt);
-  ssize_t mac_len = nflog_get_msg_packet_hwhdrlen(nfa);
+  ssize_t mac_len = 0;
   struct pcap_callback_data *cb_data = p;

+  if (nflog_get_hwtype(nfa) == DLT_EN10MB)
+    mac_len = nflog_get_msg_packet_hwhdrlen(nfa);
+
   /* Check we can handle this packet */
   switch (nfmsg->nfgen_family) {
   case AF_INET: break;
```

I will send you a trace and dump later

**Comment by paololucente** [comment]
Hi @aleksandrgilfanov , 

That patch makes sense actually. Yeah, if you can send me a dump i can double-check it working. Please, please: contribute the above patch with a Pull Request so that you get full attribution for the fix.

Paolo

**Comment by aleksandrgilfanov** [comment]
Here is dump:
[out.tar.gz](https://github.com/pmacct/pmacct/files/4516511/out.tar.gz)
And pull request:
https://github.com/pmacct/pmacct/pull/388

**Comment by paololucente** [comment]
Merged PR #388 . Thank you again @aleksandrgilfanov . Closing this now.
